### DISCUSSION TITLE
"Wer mist, mist Mist" (german: Who, who measure something, get unexpected results." (SFW translation... ;) )

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#114] [General]
Hi Travis,

Just for the fun of it, I wrote a loop, which does nothing more than toggiliing a GPIO-pin via using the related XOR-register.
Then I used my voltmeter to measure the frequency at that pin ...and got....2.14 MHz. The loop was compiled to RAM.

I used ms-delays in the loop or short calculations and all measured frequency were as rock stable as well in a
reasonable relation to each other.

The voltmeter is not the typical cheap home-improvement-type-of-thing, but it is not a frequency meter per se.

But.....2.14 MHz? Despite being happy about the speed of zeptoforth...that is a little bit too high, isn't it?

Cheers!
Tuxic

**Comment by tabemann** [comment]
Hey Tuxic,

I made a simple test program which does just this:

```
: toggle-forever  ok
  2 pin::output-pin  ok
  [ 2 bit ] literal  ok
  gpio::GPIO_OUT_XOR  ok
  begin 2dup ! again  ok
;  ok
see toggle-forever 
20031600 B500:      toggle-forever:       PUSH {LR}
20031602 3F04:                            SUBS R7, #$4
20031604 603E:                            STR R6, [R7]
20031606 2602:                            MOVS R6, #$2
20031608 4908:                            LDR R1, #$2003162C @ 10026D3D
2003160A 4788:                            BLX R1
2003160C 3F04:                            SUBS R7, #$4
2003160E 603E:                            STR R6, [R7]
20031610 2604:                            MOVS R6, #$4
20031612 3F04:                            SUBS R7, #$4
20031614 603E:                            STR R6, [R7]
20031616 4E06:                            LDR R6, #$20031630 @ D000001C
20031618 6838:      1:                    LDR R0, [R7]
2003161A 3F08:                            SUBS R7, #$8
2003161C 607E:                            STR R6, [R7, #$4]
2003161E 6038:                            STR R0, [R7]
20031620 CF01:                            LDMIA R7!, {R0}
20031622 6030:                            STR R0, [R6]
20031624 CF40:                            LDMIA R7!, {R6}
20031626 E7F7:                            B 1B <$20031618>
20031628 BD00:                            POP {PC}
```

The inner loop of this is eight instructions, and zeptoforth by default runs on the RP2040 at 125 MHz. I ran this code on a Pico W, and it generated a square wave at 3.91 MHz according to my logic analyzer (which was set to 1 GHz). With optimized assembly or a PIO I have been able to make it go much faster than this.

Travis

**Comment by mcctuxic** [comment]
Hi Travis,

an absolute positive "OH!" from my side here...
That are frequencies, which I might be able to receive
with my little shortwave receiver!!!! Incredible!
I never had expected such switching speeds! :)

For me your Forth-Fu is - hrrrrmmm - advanced...at least..., Travis :)

Let's see...

\ Start defining the word 'toggle-forever'
: toggle-forever
\ set pin 2 as an output pin
  2 pin::output-pin
\ jump back to the REPL, enter '2 bit <return>', grab the result 
\ delete '[ 2 bit ]' from the code and insert the result '4' instead.
\ than compile '4' as a literal. 
  [ 2 bit ] literal
\ put the address of the Gpio Output Register onto the stack
  gpio::GPIO_OUT_XOR  
\ start an endless loop: 2duplify the address ... and here I got lost:
\ before 2dup the stack consists of
\ 4 <gpiooutxor-addr 2 cells>
\ after 2dup the it looks like this:
\ 4 <gpiooutxor-addr 2 cells> <gpiooutxor-addr 2 cells>
\ with '!' the address is written onto .... itsself?
\ no...I am sure wrong here...but I can't see what I have
\ mistaken...
\
\ : help ask_travis ;
\
  begin 2dup ! again 

Cheers!
frequency-flabbergasted Tuxic

**Comment by tabemann** [comment]
On Sun, Jul 14, 2024, 00:18 mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> an absolute positive "OH!" from my side here...
> That are frequencies, which I might be able to receive
> with my little shortwave receiver!!!! Incredible!
> I never had expected such switching speeds! :)
>
> For me your Forth-Fu is - hrrrrmmm - advanced...at least..., Travis :)
>
> Let's see...
>
> \ Start defining the word 'toggle-forever'
> : toggle-forever
> \ set pin 2 as an output pin
> 2 pin::output-pin
> \ jump back to the REPL, enter '2 bit <return>', grab the result
> \ delete '[ 2 bit ]' from the code and insert the result '4' instead.
> \ than compile '4' as a literal.
> [ 2 bit ] literal
> \ put the address of the Gpio Output Register onto the stack
> gpio::GPIO_OUT_XOR
> \ start an endless loop: 2duplify the address ... and here I got lost:
> \ before 2dup the stack consists of
> \ 4 <gpiooutxor-addr 2 cells>
> \ after 2dup the it looks like this:
> \ 4 <gpiooutxor-addr 2 cells> <gpiooutxor-addr 2 cells>
> \ with '!' the address is written onto .... itsself?
> \ no...I am sure wrong here...but I can't see what I have
> \ mistaken...
> \
> \ : help ask_travis ;
> \
> begin 2dup ! again
>
> Cheers!
> frequency-flabbergasted Tuxic
>

2DUP here duplicates both the 4 (for GPIO 2) and the address pushed by
gpio::GPIO_OUT_XOR, the GPIO output register XOR alias, on the data stack
and then ! pops both off the stack, storing the former at the latter, hence
toggling GPIO 2.

On 07/13 04:45, tabemann wrote:
> > Hey Tuxic,
> >
> > I made a simple test program which does just this:
> >
> > ```
> > : toggle-forever ok
> > 2 pin::output-pin ok
> > [ 2 bit ] literal ok
> > gpio::GPIO_OUT_XOR ok
> > begin 2dup ! again ok
> > ; ok
> > see toggle-forever
> > 20031600 B500: toggle-forever: PUSH {LR}
> > 20031602 3F04: SUBS R7, #$4
> > 20031604 603E: STR R6, [R7]
> > 20031606 2602: MOVS R6, #$2
> > 20031608 4908: LDR R1, #$2003162C @ 10026D3D
> > 2003160A 4788: BLX R1
> > 2003160C 3F04: SUBS R7, #$4
> > 2003160E 603E: STR R6, [R7]
> > 20031610 2604: MOVS R6, #$4
> > 20031612 3F04: SUBS R7, #$4
> > 20031614 603E: STR R6, [R7]
> > 20031616 4E06: LDR R6, #$20031630 @ D000001C
> > 20031618 6838: 1: LDR R0, [R7]
> > 2003161A 3F08: SUBS R7, #$8
> > 2003161C 607E: STR R6, [R7, #$4]
> > 2003161E 6038: STR R0, [R7]
> > 20031620 CF01: LDMIA R7!, {R0}
> > 20031622 6030: STR R0, [R6]
> > 20031624 CF40: LDMIA R7!, {R6}
> > 20031626 E7F7: B 1B <$20031618>
> > 20031628 BD00: POP {PC}
> > ```
> >
> > The inner loop of this is eight instructions, and zeptoforth by default
> runs on the RP2040 at 125 MHz. I ran this code on a Pico W, and it
> generated a square wave at 3.91 MHz according to my logic analyzer (which
> was set to 1 GHz). With optimized assembly or a PIO I have been able to
> make it go much faster than this.
>

Travis

>

**Comment by mcctuxic** [comment]
On 07/14 07:17, tabemann wrote:
> On Sun, Jul 14, 2024, 00:18 mcctuxic ***@***.***> wrote:
> 
> > Hi Travis,
> >
> > an absolute positive "OH!" from my side here...
> > That are frequencies, which I might be able to receive
> > with my little shortwave receiver!!!! Incredible!
> > I never had expected such switching speeds! :)
> >
> > For me your Forth-Fu is - hrrrrmmm - advanced...at least..., Travis :)
> >
> > Let's see...
> >
> > \ Start defining the word 'toggle-forever'
> > : toggle-forever
> > \ set pin 2 as an output pin
> > 2 pin::output-pin
> > \ jump back to the REPL, enter '2 bit <return>', grab the result
> > \ delete '[ 2 bit ]' from the code and insert the result '4' instead.
> > \ than compile '4' as a literal.
> > [ 2 bit ] literal
> > \ put the address of the Gpio Output Register onto the stack
> > gpio::GPIO_OUT_XOR
> > \ start an endless loop: 2duplify the address ... and here I got lost:
> > \ before 2dup the stack consists of
> > \ 4 <gpiooutxor-addr 2 cells>
> > \ after 2dup the it looks like this:
> > \ 4 <gpiooutxor-addr 2 cells> <gpiooutxor-addr 2 cells>
> > \ with '!' the address is written onto .... itsself?
> > \ no...I am sure wrong here...but I can't see what I have
> > \ mistaken...
> > \
> > \ : help ask_travis ;
> > \
> > begin 2dup ! again
> >
> > Cheers!
> > frequency-flabbergasted Tuxic
> >
> 
> 2DUP here duplicates both the 4 (for GPIO 2) and the address pushed by
> gpio::GPIO_OUT_XOR, the GPIO output register XOR alias, on the data stack
> and then ! pops both off the stack, storing the former at the latter, hence
> toggling GPIO 2.

...and I knew it! I knew 2DUP and what it does. But it didn't make
'click' in my head while reading the code...sorry :)

Currently I am looking at rosettacode.org especially for programming
tasks involving numbers...factorizing, finding primes,
perfect/abundant/deficit numbers and such.
Just to get more used to forth/zeptoforth.

Having a battery powered all-in-one complete development environment,
which fits into my pocket is just too neat :)
Very charming! 8)

**Comment by tabemann** [comment]
On Sun, Jul 14, 2024 at 11:29 AM mcctuxic ***@***.***> wrote:

> > 2DUP here duplicates both the 4 (for GPIO 2) and the address pushed by
> > gpio::GPIO_OUT_XOR, the GPIO output register XOR alias, on the data stack
> > and then ! pops both off the stack, storing the former at the latter,
> hence
> > toggling GPIO 2.
>
> ...and I knew it! I knew 2DUP and what it does. But it didn't make
> 'click' in my head while reading the code...sorry :)
>
> Currently I am looking at rosettacode.org especially for programming
> tasks involving numbers...factorizing, finding primes,
> perfect/abundant/deficit numbers and such.
> Just to get more used to forth/zeptoforth.
>
> Having a battery powered all-in-one complete development environment,
> which fits into my pocket is just too neat :)
> Very charming! 8)


Good luck with your zeptoforth adventures!

Travis
